### PR TITLE
Serialization improvements: final step

### PR DESCRIPTION
**Pull request by sipa**
This is the final step 🥳 of the serialization improvements extracted from #10785.

It converts the LimitedString wrapper to a new-style formatter, and updates the wallet and Qt code to use the new serialization framework. Finally all remaining old primitives are removed.

### CODE REVIEW

**Review by maflcko — COMMENTED** [review]
ACK f57088302fa3f2bf1d84c7d90abfaeaeb2a25684 code looks correct, but documentation got corrupted for some reason 🚖

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK f57088302fa3f2bf1d84c7d90abfaeaeb2a25684 code looks correct, but documentation got corrupted for some reason 🚖
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUg98gwAoENgW+nEfnRVpU+BYSwAmkWsGXmnb3mhgzZ5iaW6aADpwJ6Ja9na+N6C
xXR/Ehg7uz3Eorj7NzJSBUWBgH/Tw0lCI6v0mXHXUnbcsNS0Xaj1UGBMtI2qFYyX
PKuM9/T8SsbxpnE4QUPXoEytiS0u1sdRkGqJ7/N3TLkhg3a0ZcSeF+GEH2SV2xxB
L3njvthwIv4YAn5iCzGNyRV1/DVL4bUdk7JTjiaZUJj1GEr7BI/j+tryWahkuB54
yIGL8rQojO8vWOwVUj7AQuR7fEgpoD8LAtdsD8Oxq581+sIH3ahyl8ZeqxYDk2X/
4aej8GK+RQ1Tqb3Gc9l9/y7/XL42DoCmGMSuW6IFD17XQDTv2UV95+DlR9p4NV1w
hHX0cq0JsXaJNBQzR2CMAx+fIKs5BevHtEODENnOd0Co52XZYVmwZfUkpCOthDKN
94hAA2tQ/gP3Fo+60ub3ZAdZ2CrkSKXmL/R0Ro6ykRsLeRrN70ZpzA5yokp7vaQG
jMLT8uSs
=kX23
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3225b42876eec8e47ecaa04d904d9008dc762f25068d2120bfd47040bd83f701  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083225b42876eec8e47ecaa04d904d9008dc762f25068d2120bfd47040bd83f701f010a84bb71d0702c3142ab34c4559b4750e08fff010061f34a7b0e866ea7b48f740500bc2e608f1045ec57f1bf008807afc2403ab959e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010080386c1bdfce9417dc4f9b35bb9588c08f120d22c1b974a34301da4155615180073294d0b35b9734bff55e186330a679f3cd008f1045ec57f1bf008536d1e60327628390083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01012aaca26bdca369b20dc9952c367a63208f120f3a23e88ce5033396fec21aad81cbf0369bfaae90202835848272f6aba6247fc08f1045ec57f1af00817a46180c447c7ae0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f01050e24069407de391e85c9ae3589c06e008f1045ec57f1af008eb484a7910ea134c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Comment by MarcoFalke** [comment]
The fuzz tests are complaining:
```
test/fuzz/string.cpp:111:15: error: no template named 'LimitedString'

        const LimitedString<10> limited_string = LIMITED_STRING(random_string_1, 10);

**Comment by practicalswift** [comment]
Concept ACK

The new serialization code is clearer and thus much easier to reason about.

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -247,15 +247,13 @@ struct COutputEntry
 class CMerkleTx
 {
 public:
-    template<typename Stream>
-    void Unserialize(Stream& s)
+    SERIALIZE_METHODS(CMerkleTx, obj)
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/wallet/wallet.h:250]: You're right. I think this code was originally written before the Serialize method was removed here, and through rebases now reintroduces it. I'm happy to remove it, but perhaps it's desirable for consistency reasons.

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/wallet/wallet.h:250]: The serialization code has been removed in 05b56d1c937b7667ad51400d2f9fb674af72953f, just for reference.

[the remark this answers is not in the document]
Comment (sipa) [reply src/wallet/wallet.h]: Ok, removed here. No need to effectively revert that change.

**Comment by MarcoFalke** [comment]
re-ACK 81caf8a869 only changes since previous review 🏺

* Change author email address in one commit
* Restore uncorrupted documentation for m_pre_split
* Remove one overly eager hunk in CMerkleTx to minimize the diff
* Make the fuzz tests compile again

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 81caf8a869 only changes since previous review 🏺

* Change author email address in one commit
* Restore uncorrupted documentation for m_pre_split
* Remove one overly eager hunk in CMerkleTx to minimize the diff
* Make the fuzz tests compile again
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiyqQv/WKvoztimli1L/+qEb2zHJmcLmfgAQoCzC3GOh9Hi2Px9fK4Zsp691OsA
2T937ZSmwdWHpbVCzW+mDAQnGX9qN6h6ysISSsu3jUOWtBwCkHHUmAiZIItrDrsQ
Al44fOPEDW84S1RMm1cL71Sn+rt4qWQCAJ028cx37Wpp2zhGCBjBvZx+IY1BKJ4o
8L1TdkhoRSpDz6KFS5vpFGzePxwBl3LVFP/Q5Fx8Ijz0qFXn7Ewg8NtMXtuW8D0I
V32JuWT5qo4qFO0/lBdNrb8icN7tpfkH/VVZ9nKMp3RXiTdsGvJeC4iwLiEzOPJU
opQNWpkpJ79we9lS9Hw35BLMgDz9jRpcckImYceM46HrDrDG7kyKV+iCDsn4Edyj
U5JY835uzcKDXxWXyW6GijCX6why4qKwFUNdmB3HIpa269IaOy9cjaM4ZpGzzK1R
BEVlwsY9eo8YyOssR3gscZoBHMsJrTqId5oU6q9PodfewntD0O4h579bD1181iZ+
Tkr4vdOA
=IsD3
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `f7b948dcd680ca3ad0d8ba39d18536d69a5d93ae4b678d5c0fb27d54153de077  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108f7b948dcd680ca3ad0d8ba39d18536d69a5d93ae4b678d5c0fb27d54153de077f010fd7bee5dfbc8aa2e621cae9dd375cec908fff0100128ca79bc7bf1553597796ca84a619608f1045eca6b8df008234bc8e5dc4001800083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010325cf0ccbe545d9bd54d6aaaa6eea39b08f1045eca6b8df008eb8927bc0292c9510083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105eb030e6f22bb1a07e1c4c0be42bd15008f1045eca6b8ef0080da53cb3707b42b40083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df0108c69e6dc6682c1a643c732ae0ce4349608f1045eca6b8ef00886b456e93aca3c110083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: src/wallet/scriptpubkeyman.h [hunk 2]
```diff
@@ -110,36 +110,37 @@ class CKeyPool
...
+               (this will be the case for any wallet before the HD chain split version) */
+            fInternal = false;
         }
-        else {
-            READWRITE(fInternal);
-            READWRITE(m_pre_split);
+        try {
+            s >> m_pre_split;
+        } catch (std::ios_base::failure&) {
+            /* flag as external address if we can't read the m_pre_split boolean
```
Comment (MarcoFalke) [inline src/wallet/scriptpubkeyman.h:141]:
```suggestion
            /* flag as postsplit address if we can't read the m_pre_split boolean
```

I think this is still wrong

Comment (jonatack) [inline src/wallet/scriptpubkeyman.h:141]: 309a653 s/external/postsplit/?

```diff
-            /* flag as external address if we can't read the m_pre_split boolean
+            /* flag as postsplit address if we can't read the m_pre_split boolean
```

**Review by jonatack — COMMENTED** [review]
ACK 81caf8a869 modulo doc nit below. Review, build/tests, ran bitcoind/bitcoin-qt, fuzz build/ran string fuzzer.

```
$ src/test/fuzz/string ../qa-assets/fuzz_seed_corpus/
INFO: Seed: 1884355399
INFO: Loaded 1 modules   (499191 inline 8-bit counters): 499191 [0x562be9805588, 0x562be987f37f), 
INFO: Loaded 1 PC tables (499191 PCs): 499191 [0x562be987f380,0x562bea01d2f0), 
INFO:    52918 files found in ../qa-assets/fuzz_seed_corpus/
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 52918 min: 1b max: 3984182b total: 495166672b rss: 163Mb
#4096	pulse  cov: 5223 ft: 12262 corp: 216/949b exec/s: 2048 rss: 255Mb
#8192	pulse  cov: 5391 ft: 13542 corp: 299/1855b exec/s: 2048 rss: 342Mb

.../...

#376078	REDUCE cov: 5594 ft: 17542 corp: 576/19Kb exec/s: 1047 rss: 1123Mb L: 31/412 MS: 1 EraseBytes-
#382183	REDUCE cov: 5594 ft: 17542 corp: 576/19Kb exec/s: 1047 rss: 1123Mb L: 7/412 MS: 1 EraseBytes-
```

File: src/wallet/scriptpubkeyman.h (hunk 2 above)
Comment (sipa) [reply src/wallet/scriptpubkeyman.h]: Fixed.

Comment (sipa) [reply src/wallet/scriptpubkeyman.h]: Fixed.

File: src/wallet/scriptpubkeyman.h [hunk 3]
```diff
@@ -110,36 +110,37 @@ class CKeyPool
...
+        }
+        s << nTime << vchPubKey << fInternal << m_pre_split;
+    }
 
-    template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action) {
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
         int nVersion = s.GetVersion();
```
Comment (jonatack) [inline src/wallet/scriptpubkeyman.h:126]: ef17c03 is this new change from 309a653 intentional (needed for deserializing)?
```diff
-        int nVersion;
+        int nVersion = s.GetVersion();
```

**Comment by jonatack** [comment]
ACK 71f016c6eb42e1ac2 modulo my question at https://github.com/bitcoin/bitcoin/pull/19032#discussion_r429832792. Retested/refuzzed.

<details><summary>changes since 81caf8a: initializing <code>nVersion</code> with <code>GetVersion()</code> in <code>ScriptPubKeyMan::Unserialize()</code> and documentation fixup</summary><p>

```diff

diff --git a/src/wallet/scriptpubkeyman.h b/src/wallet/scriptpubkeyman.h
--- a/src/wallet/scriptpubkeyman.h
+++ b/src/wallet/scriptpubkeyman.h

@@ -123,7 +123,7 @@ public:
     void Unserialize(Stream& s)
     {
-        int nVersion;
+        int nVersion = s.GetVersion();
         if (!(s.GetType() & SER_GETHASH)) {
             s >> nVersion;
         }

@@ -138,8 +138,8 @@ public:
         } catch (std::ios_base::failure&) {
-            /* flag as external address if we can't read the m_pre_split boolean
-               (this will be the case for any wallet that upgrades to HD chain split)*/
+            /* flag as postsplit address if we can't read the m_pre_split boolean
+               (this will be the case for any wallet that upgrades to HD chain split) */
             m_pre_split = false;
```
</p></details>

File: src/wallet/scriptpubkeyman.h (hunk 3 above)
Comment (sipa) [reply src/wallet/scriptpubkeyman.h:126]: I doubt it's needed, as I don't think this class is ever deserialized with SER_GETHASH. However, if it were ever invoked that way, the old code was UB (nVersion would be used without being initialized).

Comment (jonatack) [reply src/wallet/scriptpubkeyman.h:126]: Thank you. Good catch and reminder.

**Comment by laanwj** [comment]
Code review ACK 71f016c6eb42e1ac2c905e04ba4d20c2009e533f

**Comment by MarcoFalke** [comment]
re-ACK 71f016c 🤴

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 71f016c 🤴
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiCcgv9FZxj/YquESruZvS2FcLiLs55nDP0DClnbFBcYts6JsuqkeMwNIiBY1Oa
A8Qx6zv8qN1PdFVXsOhzoA86abTQMGxzJkBHRqC73eDtB9n9DsSZygz+bWP+aLKN
cC7r5YRwdBmnIv/47+rELLK8DD31geVNyISZUvsPk/JfJAwJhf2uSi1MEAI2qjds
FV890WMLshTbj86n2cMCLWlN6LXL9Okkiw4SKs/qo9OzLvCXQwJzycL4grs9lHTj
M3MzZJ6W66jp1BZe9/ob3LeHGwd4qL+5gvaO3FM1y7LoxNCZAnZk7kC83nCv4TOC
MLX2aruo2I1KpumZWoICSkqqp1MXMG6qPnfu7RpntXu/0ky6bB40eyg8q6HaWQPT
jIO7F8DkEoow0jC9XNQmG4DJYNq3FhAkDAeKCxnnUsR9zSLEa2HyZ3N/VHp3THsK
hBNzdzfAQ0sAN88Gz1WFL/ky1fy4v2ynBz5hixctLGvvU7EkmYdC4QJnRs6yErvu
1X0GS7hd
=cvua
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3646ddaf3ced2728b4c9185e8341178d02575e344a8f4292869f74ec2b3f5a97  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083646ddaf3ced2728b4c9185e8341178d02575e344a8f4292869f74ec2b3f5a97f010eb2327481508399582bee1158ce613a908fff010304ecd932831040e30e161bbe9ca4e2b08f1045ede4b45f0081d08e44f1a61f9790083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0109a96b026c6ca9549e2f5fbd0a1120e6c08f1208a6c4242a55b1745e37d631d3abb132f603c230deb6685b6438c730fb2db5f4408f1045ede4b44f008a77e3efe71b01a0a0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010cd902f6a788a90855094d432fa21ac2308f1045ede4b45f008c085b5714483cb300083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f183ba0eda8942a659532322f4f7cc9408f1045ede4b45f0085acd2e7fcd19ec330083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Pull request merged**